tests/unit_tests/components/config.cpp

```diff
@@ -0,0 +1,143 @@
+#include <iostream>
+
+#include <unistd.h> // unlink, rmdir
+
+#include "common.hpp"
+#include "errors.hpp"
+#include "components/config.hpp"
+#include "components/logger.hpp"
+#include "utils/file.hpp"
+
+#include "common/test.hpp"
+
+using namespace polybar;
+
+namespace {
+
+  class temp_directory {
+  public:
+    temp_directory(string directory) {
+      if (file_util::is_absolute(directory)) {
+        m_directory = directory;
+      } else {
+        m_directory = "/tmp/" + directory;
+      }
+
+      if (m_directory.length() < 6 || m_directory.compare(m_directory.length() - 6, 6, "XXXXXX") != 0) {
+        m_directory += "_XXXXXX";
+      }
+
+      if(mkdtemp(&m_directory[0]) == nullptr) {
+        throw system_error("mkdtemp failed");
+      }
+    }
+
+    ~temp_directory() {
+      // this will fail if directory is not empty: tests should clear their own files
+      rmdir(m_directory.c_str());
+    }
+
+    const string& directory() const {
+      return m_directory;
+    }
+
+  private:
+    string m_directory;
+  };
+
+  class temp_filestream : public fd_stream<std::iostream> {
+  public:
+    explicit temp_filestream(string directory)
+      : temp_filestream(directory, "filename_XXXXXX")
+    {}
+
+    ~temp_filestream() {
+      unlink(m_full_path.c_str());
+    }
+
+    const string& filename() const {
+      return m_filename;
+    }
+
+    const string& directory() const {
+      return m_directory;
+    }
+
+    const string& full_path() const {
+      return m_full_path;
+    }
+
+  private:
+    temp_filestream(string directory, string filename)
+      : fd_stream<std::iostream>(open_temp_file(directory, filename))
+      , m_directory(std::move(directory))
+      , m_filename(std::move(filename))
+      , m_full_path(m_directory + "/" + m_filename)
+    {}
+
+    static int open_temp_file(string& directory, string& filename) {
```

Comment:
Doesn't this modify the arguments of the `temp_filestream` constructor because we use references here?

Reply:
Yes, but the arguments are passed by value and then moved in to the data member. That ensures that the `m_directory` member has the real path name of the temp directory, not the template passed to `mkstemp`. (This is private and only called in the constructor)